### ISSUE TITLE
Implement CEM/MPC gameplay action planning with the world model

### ISSUE THREAD

**Issue by Sino-Huang** [Sino-Huang/NovPhy#56]
## Problem statement

The existing BG-NS-JEPA controller selects a prediction horizon/description pair after an intervention has been supplied. It does not generate Angry Birds shots. The repository has no explicit gameplay planner that searches slingshot drag vectors, tap timing, or multi-shot action sequences with the action-conditioned world model.

This is a separate downstream gameplay workstream. It must not become a blocker for the prediction/controller research chain in #1, which explicitly excludes planning and task-success claims.

## Scope

Implement a goal-conditioned gameplay action planner that:

- represents legal single-shot actions in the game interface, including slingshot drag and tap timing;
- samples candidate action sequences with the Cross-Entropy Method (CEM);
- evaluates candidates through the frozen action-conditioned world model and declared gameplay cost;
- updates the candidate distribution for a fixed, reproducible number of CEM iterations;
- runs CEM inside receding-horizon model-predictive control (MPC): execute only the first selected shot, observe the real post-shot state, and replan;
- records requested/effective prediction horizons, accumulated rollout error diagnostics, invalid candidates, model/planner failures, compute, and wall-clock time;
- preserves the distinction between the prediction-pair controller and the gameplay action planner.

The gameplay cost must explicitly bind goal progress, terminal success/failure semantics, legal-action constraints, physical/rollout penalties, and compute terms. Candidate evaluation must use deployment-available observations and the frozen selected world-model/controller configuration.

## Acceptance criteria

- Legal slingshot drag vectors and tap times are generated and round-trip through the existing game-action interface.
- Seeded CEM produces reproducible candidate batches, elite selection, distribution updates, and selected first actions.
- Candidate action sequences are evaluated by action-conditioned model rollouts; no future engine truth is used during planning.
- The MPC loop executes one shot, consumes the resulting real observation, and replans until success, terminal failure, or a declared shot/compute limit.
- Open-loop CEM and receding-horizon MPC are distinguishable in artifacts, and recursive rollout error is measured against executed transitions.
- Planner compute and failure accounting include candidate generation, every model rollout, goal/cost evaluation, invalid candidates, replanning, and game-interface failures.
- Deterministic artifacts bind code revision, model/controller checkpoint, observation/parser configuration, goal/cost version, action bounds, CEM/MPC configuration, seed, and level identity.
- At least random-legal-action and declared heuristic/no-model planning baselines can be invoked through the same planner interface.
- No sealed gameplay evaluation outcomes are used to tune the planner.

## Dependencies and placement

Blocked by:

- #15 — Run the confirmatory oracle-symbol joint-controller experiment
- #17 — Run the frozen-encoder visual predicate stress test

This ticket is a new downstream branch after the action-conditioned predictor and deployment observation path are frozen. It does **not** block #34, #15, #16, #17, #35, #36, #37, or #38.

## Testing and evidence

- Unit tests cover legal action parameterization, seeded CEM updates, elite handling, cost aggregation, first-action execution, and stopping rules.
- Integration tests use deterministic action-conditioned rollout fixtures and verify open-loop versus replanning behavior.
- A bounded live-Unity smoke test demonstrates one complete observe-plan-act-observe-replan cycle with progress logs.
- A compact source-bound evidence artifact records configuration, identities, action traces, candidate/elite summaries, compute/failure accounting, and rerun commands.

## Non-goals

- Redefining or retraining the BG-NS-JEPA prediction model or horizon/description controller.
- Treating the prediction-pair controller as a shot selector.
- Claiming end-to-end gameplay success; that is evaluated in the separate gameplay-success ticket.
- End-to-end reinforcement learning or planner-driven fine-tuning by default.
- Using final-evaluation levels for planner tuning.

**Comment by Sino-Huang** [comment]
Roadmap placement correction: the earlier statement that this ticket does not block #35–#38 is superseded. CEM/MPC planning is now required downstream research evidence for demonstrating whether the granularity-based world model improves real action selection. The required chain is `#17 → #56 → #57 → #35 → #36 → #37 → #38`. This does not redefine the model-level hypothesis, but it does make completion of the planner and its gameplay evaluation necessary before final statistical consolidation, archiving, reporting, and terminal audit.

**Comment by Sino-Huang** [comment]
## Planner handoff: unreliable `structure-unstable` signal

Issue #17 is closed under its frozen downstream non-degradation rule, but the deployed visual parser is not a reliable `structure-unstable` detector. On final data it predicted unstable for all 1,604 available states: 14 true positives and 1,590 false positives (precision `0.0087`, F1 `0.0173`). Training contained only 28 positive frames out of 2,150, and the target depends on temporal support/state changes while the parser sees one RGB frame at a time.

For #56:

- do not treat this binary output as a validated safety or gameplay-cost oracle;
- bind and log the raw probability, thresholded value, and whether/how it changes candidate costs or selected actions;
- keep parser-disabled and oracle-symbol diagnostics distinguishable from the deployment configuration where feasible, so #57/#35 can attribute planner behavior;
- do not retrain the parser or use sealed gameplay outcomes to tune it inside #56.

This limitation does not currently block #56 because #17 measured zero controller disagreement on its six final rollouts. It must remain visible so a later gameplay failure or success is not incorrectly attributed to a reliable visual instability signal.

**Comment by Sino-Huang** [comment]
### Wayfinding for fresh sessions

The governing data-adequacy handoff is recorded in #18: https://github.com/Sino-Huang/NovPhy/issues/18#issuecomment-5447228460. Read it together with the #17 closing evidence before using visual macro predicates. It distinguishes the valid frozen cohort from the additional balance, independent-event, lineage-diversity, and temporal-observation requirements of any future parser-training successor. This reference does not add #18 as a blocker for #56.

**Comment by Sino-Huang** [comment]
## Completed: CEM/MPC gameplay planner and live Unity smoke

Implementation is complete at 4e11130a1e220f0d511db7d8b2c920ed6c00fb6b. Compact evidence is published at a98b71b67e7c92375052160139742a9e31cd0fc0 in data/runtime_evidence/issue-56/cohort-v2-gameplay-planning-smoke-summary.json.

Delivered:
- legal slingshot-relative drag/tap actions that round-trip through the existing prepared-shot interface;
- seeded CEM with reproducible batches, elite selection, distribution updates, invalid-candidate accounting, and selected first actions;
- action-conditioned frozen issue-15 model rollouts driven by the distinct frozen joint prediction-pair controller;
- open-loop and receding-horizon MPC modes plus random-legal and heuristic/no-model baselines behind one planner interface;
- explicit goal, terminal, legality, physical, rollout, compute, and zero-weight unreliable-instability cost terms;
- requested/effective horizons, executed-transition recursive error, candidate/model/goal counts, compute, failures, wall-clock, checkpoint/config/seed/level bindings, action traces, and exact artifact validation.

Verification:
- 76 targeted gameplay, action-interface, controller, parser, integrated-model, and live-adapter tests passed;
- the no-write frozen-stack dry-run completed 2 replans, 2 shots, and 16 model rollouts;
- the live Unity smoke ran on level 1 with visible foreground progress, 2 MPC replans, 640 candidates, 1,280 model rollouts, 640 goal evaluations, zero invalid candidates, and zero planner/game-interface failures;
- shot 1 was [-46, 80, tap 998 ms], followed by a real RGB observation and replan;
- shot 2 was [-40, 76, tap 994 ms], followed by the real LOST terminal state;
- accumulated executed-transition recursive rollout error was 0.0074881198816001415;
- structure-unstable probabilities were 0.5024513602256775 and 0.5024163722991943, thresholded true but explicitly zero-weighted and non-influential;
- exact validation passed for full artifact cohort-v2-gameplay-planning-evidence-v1:sha256:5d87ea58be7e6cd3d23b446e21cab7f96a25e45d817c71de886e47a858047ca5.

The terminal level failure is not a negative planner disposition and no gameplay-success claim is made here. This smoke establishes observe-plan-act-observe-replan and declared stopping behavior. Repeated held-out level success and shots-to-success remain owned by #57.

Rerun:
    python -u -m scripts.run_issue_56_gameplay_planner --dry-run
    DISPLAY=:10 python -u -m scripts.run_issue_56_gameplay_planner --live-smoke --start-engine --implementation-commit 4e11130a1e220f0d511db7d8b2c920ed6c00fb6b
    python -u -m scripts.run_issue_56_gameplay_planner --validate
